File: app/src/main/java/de/kettenblatt/geo/Geo.kt
package de.kettenblatt.geo

import kotlin.math.abs
import kotlin.math.asin
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.hypot
import kotlin.math.sin
import kotlin.math.sqrt

/** Geodesic helpers. Distances are metres, bearings degrees clockwise from north. */
object Geo {
    const val EARTH_RADIUS_M = 6_371_000.0

    // Metres per degree, for the local-plane projection. Accurate to well under
    // a metre at the scale of a single track segment, which is all it is used for.
    const val M_PER_DEG_LAT = 110_540.0
    const val M_PER_DEG_LON_EQUATOR = 111_320.0

    fun haversine(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val p1 = Math.toRadians(lat1)
        val p2 = Math.toRadians(lat2)
        val dp = p2 - p1
        val dl = Math.toRadians(lon2 - lon1)
        val h = sin(dp / 2) * sin(dp / 2) + cos(p1) * cos(p2) * sin(dl / 2) * sin(dl / 2)
        return 2 * EARTH_RADIUS_M * asin(sqrt(h))
    }

    fun bearing(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
        val p1 = Math.toRadians(lat1)
        val p2 = Math.toRadians(lat2)
        val dl = Math.toRadians(lon2 - lon1)
        val y = sin(dl) * cos(p2)
        val x = cos(p1) * sin(p2) - sin(p1) * cos(p2) * cos(dl)
        return (Math.toDegrees(atan2(y, x)) + 360.0) % 360.0
    }

    /** Smallest absolute angle between two bearings, 0..180. */
    fun bearingDelta(b1: Double, b2: Double): Double {
        val d = abs(b1 - b2) % 360.0
        return if (d > 180.0) 360.0 - d else d
    }
}

/** A track point's index and how far the search point was from it. */
data class NearestPoint(val index: Int, val distanceM: Double)

/**
 * Index of the point nearest to (lat, lon), searched over `[start, end)`.
 *
 * Restricting the range is what keeps a route that doubles back on itself from
 * binding a later feature to an earlier, coincident part of the track -- the
 * reference route revisits 47 coordinates, so an unrestricted search regularly
 * picks the wrong leg.
 */
fun <T> nearestPointIndex(
    points: List<T>,
    lat: Double,
    lon: Double,
    start: Int = 0,
    end: Int = points.size,
    latOf: (T) -> Double,
    lonOf: (T) -> Double,
): NearestPoint {
    val from = start.coerceAtLeast(0)
    val to = end.coerceAtMost(points.size)
    require(from < to) { "empty search range [$from, $to)" }

    val plane = LocalPlane(lat, lon)
    var bestIndex = from
    var bestSq = Double.MAX_VALUE
    for (i in from until to) {
        val x = plane.x(lonOf(points[i]))
        val y = plane.y(latOf(points[i]))
        val d2 = x * x + y * y
        if (d2 < bestSq) {
            bestIndex = i
            bestSq = d2
        }
    }
    return NearestPoint(bestIndex, sqrt(bestSq))
}

/**
 * Equirectangular projection around a reference point.
 *
 * Turns lat/lon into local metre coordinates so segment maths is plain Euclidean
 * geometry -- far cheaper than a haversine per segment, and exact enough over the
 * few hundred metres any one projection is used for.
 */
class LocalPlane(private val lat0: Double, private val lon0: Double) {
    private val mPerDegLon = Geo.M_PER_DEG_LON_EQUATOR * cos(Math.toRadians(lat0))

    fun x(lon: Double): Double = (lon - lon0) * mPerDegLon
    fun y(lat: Double): Double = (lat - lat0) * Geo.M_PER_DEG_LAT

    /** Inverse of [x]/[y], for turning a projected point back into a coordinate. */
    fun lon(x: Double): Double = lon0 + x / mPerDegLon
    fun lat(y: Double): Double = lat0 + y / Geo.M_PER_DEG_LAT
}

/** Where a point falls on a segment: [t] along it, [distanceM] away from it. */
data class Projection(val t: Double, val distanceM: Double)

/**
 * Project (px, py) onto the segment a->b in a plane.
 *
 * Zero-length segments -- which occur wherever a track doubles back on a single
 * point -- return t = 0 and the distance to `a` rather than dividing by zero.
 */
fun projectOntoSegment(
    px: Double, py: Double,
    ax: Double, ay: Double,
    bx: Double, by: Double,
): Projection {
    val dx = bx - ax
    val dy = by - ay
    val segSq = dx * dx + dy * dy
    if (segSq <= 1e-12) return Projection(0.0, hypot(px - ax, py - ay))

    val t = (((px - ax) * dx + (py - ay) * dy) / segSq).coerceIn(0.0, 1.0)
    return Projection(t, hypot(px - (ax + t * dx), py - (ay + t * dy)))
}

File: tools/navi/geo.py
"""Geodesic helpers shared by the preprocessing pipeline.

Distances are metres throughout. The equirectangular helpers mirror the Kotlin
implementation in the Android app (`geo/Geo.kt`) so that snapping behaves
identically on both sides.
"""

from __future__ import annotations

import math

EARTH_RADIUS_M = 6_371_000.0

# Metres per degree, used by the local-plane projection. Good to well under a
# metre at the scale of a single track segment, which is all we use it for.
M_PER_DEG_LAT = 110_540.0
M_PER_DEG_LON_EQUATOR = 111_320.0


def haversine(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Great-circle distance in metres."""
    p1, p2 = math.radians(lat1), math.radians(lat2)
    dp = p2 - p1
    dl = math.radians(lon2 - lon1)
    h = math.sin(dp / 2) ** 2 + math.cos(p1) * math.cos(p2) * math.sin(dl / 2) ** 2
    return 2 * EARTH_RADIUS_M * math.asin(math.sqrt(h))


def bearing(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Initial bearing from point 1 to point 2, in degrees clockwise from north."""
    p1, p2 = math.radians(lat1), math.radians(lat2)
    dl = math.radians(lon2 - lon1)
    y = math.sin(dl) * math.cos(p2)
    x = math.cos(p1) * math.sin(p2) - math.sin(p1) * math.cos(p2) * math.cos(dl)
    return math.degrees(math.atan2(y, x)) % 360.0


def bearing_delta(b1: float, b2: float) -> float:
    """Smallest absolute angle between two bearings, in degrees (0..180)."""
    d = abs(b1 - b2) % 360.0
    return 360.0 - d if d > 180.0 else d


def cumulative_distances(points) -> list[float]:
    """Running distance along a track, starting at 0. Length matches `points`."""
    cum = [0.0]
    total = 0.0
    for a, b in zip(points, points[1:]):
        total += haversine(a.lat, a.lon, b.lat, b.lon)
        cum.append(total)
    return cum


class LocalPlane:
    """Equirectangular projection around a reference latitude.

    Converts lat/lon to local metre coordinates so segment maths is plain
    Euclidean geometry. Accurate enough for the few-hundred-metre spans we
    project onto, and far cheaper than a haversine per segment.
    """

    __slots__ = ("lat0", "lon0", "m_per_deg_lon")

    def __init__(self, lat0: float, lon0: float) -> None:
        self.lat0 = lat0
        self.lon0 = lon0
        self.m_per_deg_lon = M_PER_DEG_LON_EQUATOR * math.cos(math.radians(lat0))

    def xy(self, lat: float, lon: float) -> tuple[float, float]:
        return ((lon - self.lon0) * self.m_per_deg_lon, (lat - self.lat0) * M_PER_DEG_LAT)


def point_to_segment(
    px: float, py: float, ax: float, ay: float, bx: float, by: float
) -> tuple[float, float, float]:
    """Project (px,py) onto segment a->b in a plane.

    Returns (t, distance, _) where t in [0,1] is the position along the segment
    and distance is the perpendicular distance to the projected point.
    Degenerate (zero-length) segments return t=0 and the distance to `a`.
    """
    dx, dy = bx - ax, by - ay
    seg_sq = dx * dx + dy * dy
    if seg_sq <= 1e-12:
        return 0.0, math.hypot(px - ax, py - ay), 0.0
    t = ((px - ax) * dx + (py - ay) * dy) / seg_sq
    t = 0.0 if t < 0.0 else (1.0 if t > 1.0 else t)
    cx, cy = ax + t * dx, ay + t * dy
    return t, math.hypot(px - cx, py - cy), seg_sq


def nearest_point_index(points, lat: float, lon: float, start: int = 0, end: int | None = None) -> tuple[int, float]:
    """Index of the track point nearest to (lat, lon), searched over [start, end).

    Restricting the range is what keeps a route that doubles back on itself from
    binding a later feature to an earlier, coincident part of the track.
    """
    if end is None:
        end = len(points)
    start = max(0, start)
    end = min(len(points), end)
    if start >= end:
        raise ValueError(f"empty search range [{start}, {end})")

    plane = LocalPlane(lat, lon)
    px, py = plane.xy(lat, lon)
    best_i, best_d2 = start, float("inf")
    for i in range(start, end):
        p = points[i]
        x, y = plane.xy(p.lat, p.lon)
        d2 = (x - px) ** 2 + (y - py) ** 2
        if d2 < best_d2:
            best_i, best_d2 = i, d2
    return best_i, math.sqrt(best_d2)
